WIP dynamic conv pr

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67982

Summary: TODO test using dynamic_conv with the quantization
workflow note: found a bug with quantized implementation of ConvTranspose3d
which needed to be corrected
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

HDCharles committed:
WIP dynamic conv pr

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "WIP dynamic conv pr"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp:
```diff
@@ -0,0 +1,184 @@
+    const at::Tensor& input,
+    bool reduce_range) {
+  /// is this needed?
+  auto input_contig = input.contiguous();
```
This should be channels last contiguous. fbgemm expects input to be in channels last format. Check the equivalent qconv implementation.

kimishpatel replied:
Actually if we want this to work for transposed conv as well, I think it is better to just not call `contiguous` here. For doing per tensor quantization I dont think it matters.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp:
```diff
@@ -0,0 +1,184 @@
+
+    at::Tensor out = apply_impl<false>(q_input, q_params.scale, q_params.zero_point);
+
+    return at::dequantize(out);
```
Put a comment here to the effect that we can have optimized kernel that produces fp32 output. This also holds for fbgemm.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp:
```diff
@@ -0,0 +1,184 @@
+  TORCH_FN(QConvDynamicInt8<3>::run));
+}
+
+/* is this needed?
+TORCH_LIBRARY_IMPL(_quantized, CPU, m) {
+  m.impl(TORCH_SELECTIVE_NAME("_quantized::linear_dynamic"),
+TORCH_FN(QLinearDynamicInt8<false>::run));
+}*/
+
```
Remove.

kimishpatel commented on aten/src/ATen/native/quantized/library.cpp:
```diff
@@ -99,6 +104,10 @@ TORCH_LIBRARY(quantized, m) {
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::conv_transpose1d(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> Tensor"));
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::conv_transpose2d(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> Tensor"));
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::conv_transpose3d(Tensor qx, __torch__.torch.classes.quantized.Conv3dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> Tensor"));
+  // do I need to add these 3 to _quantized?
```
What do you mean by `_quantized`?

kimishpatel commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -3661,6 +3661,26 @@ def test_embedding_bag_2d_indices(self):
+        X_fp32 = torch.randn(3, 3, 4, 4)
+        reduce_range = True
+        s, z = _calculate_dynamic_qparams(X_fp32, torch.quint8, reduce_range)
+        X_q = torch.quantize_per_tensor(X_fp32, s, z, torch.quint8)
```
I think general practice has been to quantize the tensor and then dequantize to get `X_fp32` due to loss in precision if you directly operator on original X_fp32. Check other tests in the file.

kimishpatel replied:
Also maybe have more test cases generated similar to other tests?

HDCharles replied:
okay, will do

kimishpatel commented on torch/ao/quantization/quantization_mappings.py:
```diff
@@ -42,6 +42,7 @@
     nn.Conv3d: nnq.Conv3d,
     nn.ConvTranspose1d: nnq.ConvTranspose1d,
     nn.ConvTranspose2d: nnq.ConvTranspose2d,
+    nn.ConvTranspose3d: nnq.ConvTranspose3d,  # @reviewer, any reason this wasn't here before?
```
No idea.

kimishpatel commented on torch/nn/quantized/dynamic/modules/conv.py:
```diff
@@ -0,0 +1,474 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+# import torch.nn.intrinsic as nni
+# import torch.nn.intrinsic.qat as nniqat
```
Remove

kimishpatel commented on torch/nn/quantized/dynamic/modules/conv.py:
```diff
@@ -0,0 +1,474 @@
+        stride = _pair(stride)
+        padding = _pair(padding)
+        dilation = _pair(dilation)
+        # Subclasses of _ConvNd need to call _init rather than __init__. See
+        # discussion on PR #49702
```
outdated comment? Probably from copying conv module.

kimishpatel commented on torch/nn/quantized/dynamic/modules/conv.py:
```diff
@@ -0,0 +1,474 @@
+            raise ValueError("Input shape must be `(N, C, H, W)`!")
+        if self.padding_mode != 'zeros':
+            _reversed_padding_repeated_twice = _reverse_repeat_padding(self.padding)
+            input = F.pad(input, _reversed_padding_repeated_twice,
```
Is this to pad with specific value? If so for quantized conv does this value need to be quantized as well?

kimishpatel commented on torch/nn/quantized/dynamic/modules/conv.py:
```diff
@@ -0,0 +1,474 @@
+    For details on input arguments, parameters, and implementation see
+    :class:`~torch.nn.ConvTranspose1d`.
+
+    .. note:: Currently only the QNNPACK engine is implemented.
```
This may not be true. But again probably here from copying.

kimishpatel commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -195,7 +195,7 @@ def get_qconv(cls, mod, activation_post_process, weight_post_process=None):
         if weight_post_process is None:
             weight_post_process = mod.qconfig.weight()
         weight_post_process(mod.weight)
-        act_scale, act_zp = activation_post_process.calculate_qparams()
```
Why this line is removed?

kimishpatel replied:
Also, if you leave it as is, what happens? Also do you not need from_float in dynamic conv module?

kimishpatel commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -204,6 +204,11 @@ def get_qconv(cls, mod, activation_post_process, weight_post_process=None):
                     mod.stride, mod.padding, mod.dilation, mod.groups,
                     mod.bias is not None, mod.padding_mode)
         qconv.set_weight_bias(qweight, mod.bias)
+        if isinstance(activation_post_process, torch.ao.quantization.PlaceholderObserver):
```
I see. It is added here.

kimishpatel replied:
And in what instance this condition is true?

HDCharles committed:
Update on "WIP dynamic conv pr"


Summary: TODO test using dynamic_conv with the quantization
workflow note: found a bug with quantized implementation of ConvTranspose3d
which needed to be corrected
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
